File: matlab/figures/fig03_bland_altman_pl_ds.m
function fig03_bland_altman_pl_ds()
% fig03_bland_altman_pl_ds  Bland-Altman for Omni PL and Omni DS (Fig 3).
%
%   Produces FOUR separate figures, each with ONE axis that overlays N3
%   (NYU data, blue circles) and U3 (USC data, orange squares):
%
%       fig03_BA_PL   : sub-THz PL   (27 NYU + 26 USC = 53 dots)
%       fig03_BA_DS   : sub-THz DS   (53 dots)
%       fig03_BA_PL7  : 6.75 GHz PL  (18 NYU + 17 USC = 35 dots)
%       fig03_BA_DS7  : 6.75 GHz DS  (35 dots)
%
%   Diff convention: B - A where A = NYU method (pl_nyu_sum / ds_nyu_sum)
%   and B = USC method (pl_usc_pdm / ds_usc_pdm). Bias>0 means USC
%   reports a higher value than NYU on the same data.
%
%   LAYOUT (self-contained -- no legend):
%     * Bias + +/-1.96 SD lines per side with colored text() labels on
%       the left (N3) / right (U3). Uses text() instead of yline Label
%       so the label color works on MATLAB <R2023a (yline 'LabelColor'
%       only exists in R2023a+).
%     * Corner bias/1.96-SD/n boxes upper-left (blue, N3) and
%       upper-right (orange, U3) replace the legend, so no southeast
%       box is left to overlap the -1.96 SD label.
%     * Larger fonts matched to fig04 so Fig 3 and Fig 4 render at the
%       same visual scale in the paper.
%
%   DATA SOURCE:
%       load_paper_ba_source() returns the paper-parity xlsx snapshot
%       from share/data/point_data/. Its PL/DS bias numbers match the
%       paper exactly (see its docstring). If those numbers drift from
%       the live UCA Results/ .xlsx output, migrate the source to
%       P.results_nyu_142 / P.results_usc_145 / P.results_nyu_7 /
%       P.results_usc_7 as fig04 already does for AS.

% Mirrors python/src/channel_analysis/figures/fig03_bland_altman_pl_ds.py
% Paper Fig 3.

plot_style();
P = paths();
if ~exist(P.out_dir, 'dir'), mkdir(P.out_dir); end

T = load_paper_ba_source();

render_one(T, P, 'fig03_BA_PL',  'Bland-Altman: Omni PL (sub-THz)', ...
           'subTHz', 'pl_nyu_sum', 'pl_usc_pdm', 'dB', 'PL');
render_one(T, P, 'fig03_BA_DS',  'Bland-Altman: Omni DS (sub-THz)', ...
           'subTHz', 'ds_nyu_sum', 'ds_usc_pdm', 'ns', 'DS');
render_one(T, P, 'fig03_BA_PL7', 'Bland-Altman: Omni PL (6.75 GHz)', ...
           'FR1C',   'pl_nyu_sum', 'pl_usc_pdm', 'dB', 'PL');
render_one(T, P, 'fig03_BA_DS7', 'Bland-Altman: Omni DS (6.75 GHz)', ...
           'FR1C',   'ds_nyu_sum', 'ds_usc_pdm', 'ns', 'DS');
end


% ===========================================================================
function render_one(T, P, stem, ttl, band, col_a, col_b, unit, metric_lbl)
% Build one overlaid BA figure (N3 + U3 on a single axis). Same layout
% as fig04 render_one so the two figures read at identical scale.

sub_nyu = T(T.institution == "NYU" & T.band == string(band), :);
sub_usc = T(T.institution == "USC" & T.band == string(band), :);

a_n = sub_nyu.(col_a);  b_n = sub_nyu.(col_b);
a_u = sub_usc.(col_a);  b_u = sub_usc.(col_b);

% Match Plot_BlandAltman_PL_DS_AS.m filter: finite-only.
m_n = isfinite(a_n) & isfinite(b_n);
m_u = isfinite(a_u) & isfinite(b_u);
a_n = a_n(m_n);  b_n = b_n(m_n);
a_u = a_u(m_u);  b_u = b_u(m_u);

% Diff = B - A (USC method - NYU method) per paper convention.
res_n3 = bland_altman(b_n, a_n);
res_u3 = bland_altman(b_u, a_u);

% Print so the user can cross-check against Plot_BlandAltman_PL_DS_AS.m.
fprintf('  %s:\n', stem);
fprintf('    N3 (NYU data): Bias=%+.2f %s, SD=%.2f, n(NYU)=%d\n', ...
        res_n3.bias, unit, res_n3.sd, res_n3.n);
fprintf('    U3 (USC data): Bias=%+.2f %s, SD=%.2f, n(USC)=%d\n', ...
        res_u3.bias, unit, res_u3.sd, res_u3.n);

% --- Colors: NYU blue, USC orange.
c_nyu      = [0.00 0.45 0.74];
c_nyu_face = [0.60 0.78 0.92];
c_usc      = [0.85 0.33 0.10];
c_usc_face = [0.98 0.78 0.68];

fig = figure('Position', [100 100 1200 800], 'Color', 'w');
ax  = axes('Parent', fig);
hold(ax, 'on'); grid(ax, 'on'); box(ax, 'on');
set(ax, 'FontSize', 26, 'LineWidth', 1.0);

% --- Scatter (with DisplayName so markers show in the legend box).
h_n3 = scatter(ax, res_n3.mean, res_n3.diff, 320, 'o', ...
        'MarkerFaceColor', c_nyu_face, ...
        'MarkerEdgeColor', c_nyu, 'LineWidth', 2.0, ...
        'DisplayName', sprintf('N3: NYU data (n=%d)', res_n3.n));
h_u3 = scatter(ax, res_u3.mean, res_u3.diff, 320, 's', ...
        'MarkerFaceColor', c_usc_face, ...
        'MarkerEdgeColor', c_usc, 'LineWidth', 2.0, ...
        'DisplayName', sprintf('U3: USC data (n=%d)', res_u3.n));

% --- Bias (solid) + +/-1.96 SD (dashed) lines per side.
% Keep yline handles out of the legend (HandleVisibility off) -- the
% legend below only carries the two marker symbols + dataset labels.
yline(ax, res_n3.bias,     '-',  'Color', c_nyu, 'LineWidth', 2.2, 'HandleVisibility', 'off');
yline(ax, res_n3.loa_high, '--', 'Color', c_nyu, 'LineWidth', 1.8, 'HandleVisibility', 'off');
yline(ax, res_n3.loa_low,  '--', 'Color', c_nyu, 'LineWidth', 1.8, 'HandleVisibility', 'off');
yline(ax, res_u3.bias,     '-',  'Color', c_usc, 'LineWidth', 2.2, 'HandleVisibility', 'off');
yline(ax, res_u3.loa_high, '--', 'Color', c_usc, 'LineWidth', 1.8, 'HandleVisibility', 'off');
yline(ax, res_u3.loa_low,  '--', 'Color', c_usc, 'LineWidth', 1.8, 'HandleVisibility', 'off');
yline(ax, 0, ':', 'Color', [0.5 0.5 0.5], 'LineWidth', 0.8, 'HandleVisibility', 'off');

xlabel(ax, sprintf('Mean of paired %s [%s]', metric_lbl, unit), 'FontSize', 30);
ylabel(ax, sprintf('USC - NYU [%s]', unit),                     'FontSize', 30);
title(ax,  ttl, 'FontSize', 32);

% --- Y-axis padding:
%   +40 %% top    : room for the two 3-line stat boxes above the
%                   highest line label (loa_high_norm <= 0.74).
%   +25 %% bottom : clears space for the legend inside the axes (south,
%                   horizontal). The -1.96 SD labels sit ABOVE the
%                   legend (the LoA lines land around y_norm = 0.20
%                   with this padding), so no collision.
drawnow;
yl = ylim(ax);  yspan = max(yl(2) - yl(1), eps);
ylim(ax, [yl(1) - 0.25*yspan, yl(2) + 0.40*yspan]);

% --- Colored in-axes line labels in DATA coords so color works on
% older MATLABs. N3 on the LEFT (ha='left'), U3 on the RIGHT ('right').
xl = xlim(ax);  dx = 0.02 * diff(xl);
xp_l = xl(1) + dx;   xp_r = xl(2) - dx;
fs_lbl = 24;

text(ax, xp_l, res_n3.bias,     'Bias (N3)',     'Color', c_nyu, ...
     'HorizontalAlignment', 'left',  'VerticalAlignment', 'bottom', ...
     'FontSize', fs_lbl, 'FontWeight', 'bold');
text(ax, xp_l, res_n3.loa_high, '+1.96 SD (N3)', 'Color', c_nyu, ...
     'HorizontalAlignment', 'left',  'VerticalAlignment', 'bottom', ...
     'FontSize', fs_lbl, 'FontWeight', 'bold');
text(ax, xp_l, res_n3.loa_low,  '-1.96 SD (N3)', 'Color', c_nyu, ...
     'HorizontalAlignment', 'left',  'VerticalAlignment', 'top', ...
     'FontSize', fs_lbl, 'FontWeight', 'bold');
text(ax, xp_r, res_u3.bias,     'Bias (U3)',     'Color', c_usc, ...
     'HorizontalAlignment', 'right', 'VerticalAlignment', 'bottom', ...
     'FontSize', fs_lbl, 'FontWeight', 'bold');
text(ax, xp_r, res_u3.loa_high, '+1.96 SD (U3)', 'Color', c_usc, ...
     'HorizontalAlignment', 'right', 'VerticalAlignment', 'bottom', ...
     'FontSize', fs_lbl, 'FontWeight', 'bold');
text(ax, xp_r, res_u3.loa_low,  '-1.96 SD (U3)', 'Color', c_usc, ...
     'HorizontalAlignment', 'right', 'VerticalAlignment', 'top', ...
     'FontSize', fs_lbl, 'FontWeight', 'bold');

% --- Corner stat boxes. One line per line-in-plot so each notation
% (Bias, +1.96 SD, -1.96 SD) appears as a LABEL-TEXT entry next to
% its numeric value. The +1.96 SD / -1.96 SD values shown are the
% upper/lower limits-of-agreement (bias +/- 1.96*SD), matching the
% yline positions in the plot.
text(ax, 0.02, 0.97, ...
     sprintf(['N3 Bias = %+.2f %s\n' ...
              '+1.96 SD = %+.2f %s\n' ...
              '-1.96 SD = %+.2f %s'], ...
             res_n3.bias,     unit, ...
             res_n3.loa_high, unit, ...
             res_n3.loa_low,  unit), ...
     'Units', 'normalized', 'Color', c_nyu, ...
     'HorizontalAlignment', 'left',  'VerticalAlignment', 'top', ...
     'FontSize', 22, 'FontWeight', 'bold', ...
     'BackgroundColor', 'w', 'EdgeColor', c_nyu, 'Margin', 4);
text(ax, 0.98, 0.97, ...
     sprintf(['U3 Bias = %+.2f %s\n' ...
              '+1.96 SD = %+.2f %s\n' ...
              '-1.96 SD = %+.2f %s'], ...
             res_u3.bias,     unit, ...
             res_u3.loa_high, unit, ...
             res_u3.loa_low,  unit), ...
     'Units', 'normalized', 'Color', c_usc, ...
     'HorizontalAlignment', 'right', 'VerticalAlignment', 'top', ...
     'FontSize', 22, 'FontWeight', 'bold', ...
     'BackgroundColor', 'w', 'EdgeColor', c_usc, 'Margin', 4);

% --- Legend INSIDE the axes, bottom-center, horizontal. ItemTokenSize
% enlarges the marker icon in the legend beyond MATLAB's default so
% the in-legend marker matches the visual weight of the scatter.
leg = legend(ax, [h_n3, h_u3], 'Location', 'south', ...
       'Orientation', 'horizontal', 'FontSize', 24, 'Box', 'on');
leg.ItemTokenSize = [60 36];

save_figure(fig, P.out_dir, stem, 'tight');
close(fig);
end

File: share/python/src/channel_analysis/figures/fig03_bland_altman_pl_ds.py
"""Fig. 3 - Bland-Altman for PL and Omni DS, per band.

Paper layout (from Plot_BlandAltman_PL_DS_AS.m): ONE axis per figure with
N3 (NYU data, blue circles) and U3 (USC data, orange squares) overlaid.
Difference convention: USC method - NYU method (B - A) for BOTH sides, so
positive bias means USC reports a larger value than NYU on the same data.

Four figures written:
    BA_PL        : sub-THz PL   (27 NYU + 26 USC = 53 dots)
    BA_DS        : sub-THz DS   (53 dots)
    BA_PL7       : 6.75 GHz PL  (18 NYU + 17 USC = 35 dots)
    BA_DS7       : 6.75 GHz DS  (35 dots)

Data source: ``load_paper_ba_source`` in ``channel_analysis.io`` -- reads
the four authoritative per-location xlsx tables (``nyu_142_results.xlsx``
etc.) that the paper's BA script consumed. Produces paper-identical bias
values (N3 sub-THz +1.76 dB / 1.23, U3 sub-THz +5.84 dB / 2.01, N3
6.75 GHz +3.53 dB / 2.76, U3 6.75 GHz +6.65 dB / 4.95).
"""
from __future__ import annotations

import matplotlib.pyplot as plt
import numpy as np

from .. import config
from ..bland_altman import bland_altman
from ..io import load_paper_ba_source
from ._common import apply_style, save


def _render_one(stem, title, band, metric_a, metric_b, unit, xlabel_metric):
    df = load_paper_ba_source()
    sub_nyu = df[(df.institution == "NYU") & (df.band == band)]
    sub_usc = df[(df.institution == "USC") & (df.band == band)]

    a_nyu = sub_nyu[metric_a].to_numpy()
    b_nyu = sub_nyu[metric_b].to_numpy()
    a_usc = sub_usc[metric_a].to_numpy()
    b_usc = sub_usc[metric_b].to_numpy()

    # Paper convention: diff = USC_method - NYU_method (B - A).
    res_n3 = bland_altman(b_nyu, a_nyu)
    res_u3 = bland_altman(b_usc, a_usc)
    diff_n3 = b_nyu - a_nyu;  mean_n3 = 0.5 * (a_nyu + b_nyu)
    diff_u3 = b_usc - a_usc;  mean_u3 = 0.5 * (a_usc + b_usc)

    fig, ax = plt.subplots(figsize=(12, 6))
    # Legend entries carry n / bias / SD; no free text boxes to collide.
    ax.scatter(mean_n3, diff_n3, s=120, marker="o",
               facecolor=config.COLORS["nyu_face"],
               edgecolor=config.COLORS["nyu"], linewidths=1.8,
               label=(f"N3 (NYU data, n(NYU)={res_n3.n}): "
                      f"bias={res_n3.bias:+.2f} {unit}, "
                      f"1.96 SD={1.96*res_n3.sd:.2f}"))
    ax.scatter(mean_u3, diff_u3, s=120, marker="s",
               facecolor=config.COLORS["usc_face"],
               edgecolor=config.COLORS["usc"], linewidths=1.8,
               label=(f"U3 (USC data, n(USC)={res_u3.n}): "
                      f"bias={res_u3.bias:+.2f} {unit}, "
                      f"1.96 SD={1.96*res_u3.sd:.2f}"))

    # Bias lines (solid) and +/-1.96 SD lines (dashed) per side, with
    # in-axes labels ("+1.96 SD (N3)" etc.) so the LoA interpretation is
    # visible without a second legend.
    for res, color, tag, ha in (
        (res_n3, config.COLORS["nyu"], "N3", "left"),
        (res_u3, config.COLORS["usc"], "U3", "right"),
    ):
        ax.axhline(res.bias, color=color, linewidth=2.0, linestyle="-")
        ax.axhline(res.loa_high, color=color, linewidth=1.6, linestyle="--")
        ax.axhline(res.loa_low,  color=color, linewidth=1.6, linestyle="--")
        # x=0.01 for N3 (left), x=0.99 for U3 (right) in axes coords
        x_text = 0.01 if ha == "left" else 0.99
        ax.text(x_text, res.loa_high, f" +1.96 SD ({tag}) ",
                transform=ax.get_yaxis_transform(),
                color=color, ha=ha, va="bottom", fontsize=10)
        ax.text(x_text, res.loa_low,  f" -1.96 SD ({tag}) ",
                transform=ax.get_yaxis_transform(),
                color=color, ha=ha, va="top", fontsize=10)

    ax.axhline(0, color="0.5", linewidth=0.8, linestyle=":")
    ax.set_xlabel(f"Mean of paired {xlabel_metric} [{unit}]")
    ax.set_ylabel(f"Difference (USC method - NYU method) [{unit}]")
    ax.set_title(title)

    # Put legend BELOW the plot -- lots of info in the labels, and keeping
    # it outside the data area guarantees no overlap with scatter.
    ax.legend(loc="upper center", bbox_to_anchor=(0.5, -0.18),
              ncol=1, fontsize=10, frameon=True)

    fig.tight_layout()
    save(fig, stem)
    return {
        "N3": {"bias": res_n3.bias, "sd": res_n3.sd, "n": res_n3.n,
               "loa_low": res_n3.loa_low, "loa_high": res_n3.loa_high},
        "U3": {"bias": res_u3.bias, "sd": res_u3.sd, "n": res_u3.n,
               "loa_low": res_u3.loa_low, "loa_high": res_u3.loa_high},
    }


def render() -> dict:
    apply_style()
    stats: dict = {}
    stats["BA_PL"]  = _render_one("fig03_BA_PL",  "Bland-Altman: Omni PL (sub-THz)",
                                   "subTHz", "pl_nyu_sum", "pl_usc_pdm",
                                   "dB", "PL")
    stats["BA_DS"]  = _render_one("fig03_BA_DS",  "Bland-Altman: Omni DS (sub-THz)",
                                   "subTHz", "ds_nyu_sum", "ds_usc_pdm",
                                   "ns", "DS")
    stats["BA_PL7"] = _render_one("fig03_BA_PL7", "Bland-Altman: Omni PL (6.75 GHz)",
                                   "FR1C",   "pl_nyu_sum", "pl_usc_pdm",
                                   "dB", "PL")
    stats["BA_DS7"] = _render_one("fig03_BA_DS7", "Bland-Altman: Omni DS (6.75 GHz)",
                                   "FR1C",   "ds_nyu_sum", "ds_usc_pdm",
                                   "ns", "DS")
    return stats
